### PR TITLE
Merge from `dev` to `master` for version 1.7.0

### PR DESCRIPTION
**Pull request by jpdorsch**
**This PR that merges `dev` branch to `master` includes**
- `reservations` microservice
  - created and configured to interface with `rsvmgmt` tool
  - configured Kong to accept incomming calls and redirection to `{firecrest_url}/reservations`
- CI Pipeline
  - `RESTAPI-` branches (`ci/dev` files)
    - executes `test-build.ini` pytest environment (direct calls to all endpoints without using gateway)
    - New approach using declarative Jenkins pipelines
    - Stages call bash files (`ci/dev/retest.sh`, `ci/dev/refresh.sh`, etc) instead of executing commands in Jenkinsfile itself 
    - added `tester` image to isolate and encapsulate test execution giving a unified dependencies for tests execution
    - removed `docker-compose`'s `network=host` mode, since it's deprecated for versions `>1.26`
  - `dev` branch tests (`ci/pre-prod` files)
     - executes `demo.ini` pytest environment (uses demo Keycloak, and Kong to emulate the same environment of execution)
     - Uses Jenkins pipelines connecting with AWX servers (by using `ansibleTower` commands)
     - Images are stored in firecrest registry machine and tagged with github commit ID
     - Once tests are passed, images are tagged with `firecrest-tds` 
     - Uses a provisioned machine to run tests
   - Added tests for `reservations` microservice
   - Added more files to tests `storage/xfer-internal` API
- `storage` microservice
  - SLURM account configuration for `storage/xfer-internal` API (env var `F7T_USE_SLURM_ACCOUNT`)
  - Change in the JSON format of the output for temp upload form in order to match parameters of `requests` library (such as `params`, `data`, `headers`)
  - replaced `wget` command by `curl` command (therefore parameters have been changed)
- `compute` microservice
  - Reformatted output of `DELETE compute/jobs/` in order to avoid `scancel` standard error output
  - fixed `targetPath` check in `POST compute/jobs/upload` call
- `certificator` microservice
  - checks for SSL exceptions when connecting with OPA
- Demo cluster
  - Added `rsvmgmt` tool and configured SLURM (`slurmdb.conf`)  and `sudoers.d` files to match required configuration
  - Created `f7t` SLURM feature to be used in `rsvmgmt` tool
  - Added `rsyslog` to use SSH log and filter script with `logger` command
- SSH filter and log scrip
  - now uses `logger` command
  - added `$USER` name to the log line
  - removed `wget` command, replaced by `curl` command
- API specification
  - Added `reservations` specification
  - Added `account` parameter in `storage/xfer-internal` API
  - Changed to version `1.7.0`
- SSL
  - Added SSL support for microservices and OPA
  - Added SSL certs and keys for demo environment
  - in `common.env` added `F7T_USE_SSL` variable to specify whether or not use SSL in the deployment
- Docker images
  - Created `f7t-base` image, a common image for all microservices
  - Fixed all requirements and dependencies of microservices and demo cluster, Kong, Minio, Keylcoak and OPA
- OPA: added `service-account-firecrest-test` user to be used for `client_credentials` OIDC grant type tests.
- Added `pyfirecrest` submodule
- Added `CONTRIBUTORS.md`file
- Updated license's year
- Fixed typos in `utilities` microservice.

### CODE REVIEW

File: deploy/test-build/ssl/f7t_internal.crt
Comment (taliaga) [inline deploy/test-build/ssl/f7t_internal.crt:1]: On a second thought, these may no longer be needed, if we agreed to go without ssl for the `test-build` cluster. To keep in mind. @fcruzcscs PTAL at the SSL change introduced in [test-build](https://github.com/eth-cscs/firecrest/pull/76/files#diff-df8189504f0811599327a3161d585936fabc2689aef146c9745bab847e972074)

File: doc/openapi/firecrest-api.yaml
Comment (taliaga) [inline doc/openapi/firecrest-api.yaml:12]: Big bump! Tasks suggestion for the backlog regarding the spec: 
- Integrate API version check in CICD
- Remove duplication in both specs
- Consider adding a layer like connexion or fastapi to increase coverage, consistency and validation of declared/implemented schemas, params, return types, etc.

File: src/certificator/certificator.py
Comment (taliaga) [inline src/certificator/certificator.py:33]: Should be a cast to boolean instead. Setting `F7T_USE_SSL=False` is not disabling SSL because in the code the flag is being evaluated a string (`if F7T_USE_SSL: ...`)
Either we fix it now or add a task to backlog.

File: src/common/cscs_api_common.py
Comment (taliaga) [inline src/common/cscs_api_common.py:52]: ditto

File: src/common/cscs_api_common.py
Comment (taliaga) [inline src/common/cscs_api_common.py:395]: May be safer to just log a `.format(result)`. Result may not have an `error` key (missing `else` above).

Also, logging at least the `*_errno` variables may also prove helpful in debugging.

File: src/compute/compute.py
Comment (taliaga) [inline src/compute/compute.py:40]: ditto

File: src/compute/compute.py
Comment (taliaga) [inline src/compute/compute.py:827]: error is such a common word in error messages that this risks unintentionally prunning error mesages.
I'd rather look for the specific prefix you want to cleanup (something like `scancel: error:` or something more specific like that)

Since this is not storage specific, and I think I've seen something like this before (IIRC in `cscs_api_common`), I'd consider moving this check to a more visible/reusable place.

File: src/status/status.py
Comment (taliaga) [inline src/status/status.py:37]: Ditto. I started to see a pattern, potential candidate for a refactoring :)

File: src/storage/swiftOS.py
Comment (taliaga) [inline src/storage/swiftOS.py:227]: I'm not familiar with the storage workflow yet, but I guess someone should be doing sanitization of the parameters building a curl command.

**Review by taliaga — COMMENTED** [review]
As a new reviewer of this project, I may be commenting on things that are out-of-scope or already tracked. Sorry about that.

I'm just mentioning what I see, as requested. I'll let you decide if something should be done before the merge or moved to the backlog.

File: .gitmodules
Comment (jpdorsch) [reply .gitmodules:1]: The idea is that `pyfirecrest` is also part of the repository, and that users should be able to interact with latest releases in `firecrest` main repository.
+1 to README updated.

File: ci/pre-prod/build_image_role/tasks/main.yml
Comment (jpdorsch) [reply ci/pre-prod/build_image_role/tasks/main.yml:7]: Will put it in the backlog

File: ci/pre-prod/provision_test_server.yml
Comment (jpdorsch) [reply ci/pre-prod/provision_test_server.yml:49]: Makes sense. Also on the backlog for next release.

File: ci/pre-prod/deploy_demo_playbook.yml
Comment (jpdorsch) [reply ci/pre-prod/deploy_demo_playbook.yml:22]: +1

File: ci/pre-prod/tag_image_role/vars/main.yml
Comment (jpdorsch) [reply ci/pre-prod/tag_image_role/vars/main.yml:17]: minio doesn't deploy on TDS, that's why, but the comment should be removed.

File: deploy/test-build/cluster/rsvsvc/rsvmgmt_users
Comment (jpdorsch) [reply deploy/test-build/cluster/rsvsvc/rsvmgmt_users:1]: That won't be possible, the repository from which `rsvmgmt` is taken is a private one.

File: deploy/test-build/ssl/f7t_internal.crt
Comment (jpdorsch) [reply deploy/test-build/ssl/f7t_internal.crt:1]: For `test-build` it could be surelly removed, we'll keep it until next release.

File: doc/openapi/firecrest-api.yaml
Comment (jpdorsch) [reply doc/openapi/firecrest-api.yaml:12]: The "bump" is given because in the other releases openapi version change was missing.
+1 to suggestions.

File: src/certificator/certificator.py
Comment (jpdorsch) [reply src/certificator/certificator.py:33]: We should have a new release after mid-March, so this could be redirected to issues.

File: src/compute/compute.py
Comment (jpdorsch) [reply src/compute/compute.py:827]: This is specific from `DELETE compute/jobs` so there shouldn't be another occurrence. We could be more specific on the search of the string, but since `str.index(s)` captures the first occurrence of a substring it should be enough.

File: src/status/status.py
Comment (jpdorsch) [reply src/status/status.py:37]: Yes, there is not `bool()` casting in environment variables.

File: src/storage/swiftOS.py
Comment (jpdorsch) [reply src/storage/swiftOS.py:227]: Yes, `swift_url`, `swift_version`, `swift_account` are ser via environment variables.
In the case of `containername` is taken from the user name in the token, and `prefix` from the task ID.

**Review by taliaga — APPROVED** [review]

**Review by ekouts — APPROVED** [review]

File: deploy/test-build/ssl/f7t_internal.crt
Comment (fcruzcscs) [reply deploy/test-build/ssl/f7t_internal.crt:1]: in the future we should move SSL and other security features to a service mesh, but that is future work

File: src/compute/compute.py
Comment (fcruzcscs) [inline src/compute/compute.py:486]: why is this not?

data = jsonify(description="Failed to submit job file",error=e)

however, I cannot foresee any other error other than the once in L490, no?

File: ci/pre-prod/Jenkinsfile
Comment (jpdorsch) [reply ci/pre-prod/Jenkinsfile]: the testing branch `dev` is set on AWX

File: deploy/demo/deploy_demo_playbook.yml
Comment (jpdorsch) [reply deploy/demo/deploy_demo_playbook.yml]: Between the redis:5 image and a local copy, the only thing that changes is the configuration file that is mounted as a volume to the container

File: src/compute/compute.py
Comment (jpdorsch) [reply src/compute/compute.py:486]: Since the try-except block is wrapping only L485's sentence, we considered that it is fair to state that the error is that the parameter is not set.

File: ci/pre-prod/provision_test_server.yml
Comment (jpdorsch) [reply ci/pre-prod/provision_test_server.yml:58]: I think that that is something that was left from an old commit. It shouldn't be there.

File: src/compute/compute.py
Comment (fcruzcscs) [reply src/compute/compute.py:486]: just not to catch something unexpected what about using
"except KeyError"

File: deploy/demo/deploy_demo_playbook.yml
Comment (fcruzcscs) [reply deploy/demo/deploy_demo_playbook.yml]: yeah, I noticed we did the change to use local copies, somehow, this change was reverted. Maybe it is something to lock at when we move to k8s

File: src/compute/compute.py
Comment (jpdorsch) [reply src/compute/compute.py:486]: LGTM

**Review by fcruzcscs — APPROVED** [review]

**Review by fcruzcscs — APPROVED** [review]

**Pull request merged**
